sdk/include/opentelemetry/sdk/common/http_client.h

```diff
@@ -0,0 +1,149 @@
+#pragma once
+
+#include <chrono>
+#include <string>
+#include <vector>
+#include "opentelemetry/nostd/function_ref.h"
+#include "opentelemetry/nostd/string_view.h"
+#include "opentelemetry/version.h"
+
+/*
+ Usage Example
+
+ struct SimpleReponseHandler: public ResponseHandler {
+      void OnResponse(Response& res) noexcept override
+      {
+           if (res.IsSuccess()) {
+            res.GetNextHeader([](nostd::string_view name, std::string value) -> bool {
+                std::cout << "Header Name:" << name << " Header Value:"<< value ;
+                return true;
+            });
+            .. process response body
+           }
+      }
+
+      void OnError(nostd::string_view err) noexcept override
+      {
+          std::cout << " Error:" << err;
+      }
+  };
+
+  SessionManager sessionManager; // implementer can provide singleton implementation for it
+  auto session = sessionManager.createSession("localhost", 8000);
+  auto request = session->CreateRequest();
+  request->AddHeader(..);
+  SimpleResponseHandler res_handler;
+  session->SendRequest(res_handler);
+  session->FinishSession() // optionally in the end
+  ...shutdown
+  sessionManager.FinishAllSessions()
+*/
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace common
+{
+namespace http
+{
+
+enum class Method
+{
+  Get,
+  Post,
+  Put,
+  Options,
+  Head,
+  Patch,
+  Delete
+};
+
+enum class SessionState
+{
+  Created,  // session object is created
+  Ongoing,  // http(s) request is ongoing
+  Finished  // https(s) request is finished
+};
+
+using Body       = std::vector<uint8_t>;
+using StatusCode = uint16_t;
+
+class Request
+{
+public:
+  virtual void SetMethod(Method method) noexcept = 0;
+
+  virtual void SetUri(const nostd::string_view &uri) noexcept = 0;
+
+  virtual void SetBody(Body &body) noexcept = 0;
+
+  virtual void AddHeader(const nostd::string_view &name, const std::string &value) noexcept = 0;
+
+  virtual void ReplaceHeader(const nostd::string_view &name, const std::string &value) noexcept = 0;
+
+  virtual void SetTimeoutMs(std::chrono::milliseconds timeout_ms) noexcept = 0;
+
+  virtual ~Request() = default;
+};
+
+class Response
+{
+public:
+  virtual const Body &GetBody() const noexcept = 0;
+
+  virtual bool GetNextHeader(
+      nostd::function_ref<bool(nostd::string_view name, std::string value)> callable) const
+      noexcept = 0;
+
+  virtual bool GetNextHeader(
+      const nostd::string_view &key,
+      nostd::function_ref<bool(nostd::string_view name, std::string value)> callable) const
+      noexcept = 0;
+
+  virtual StatusCode GetStatusCode() const noexcept = 0;
+
+  virtual ~Response() = default;
+};
+
+class ResponseHandler
+{
+public:
+  virtual void OnResponse(Response &) noexcept = 0;
```

Comment:
I wonder if it makes sense a wider set of events here, that is not only covering the response, but entire session - originating from connect, to the moment when we post/get, then when we already read back the response. Also having consolidated set of events / handler for both parts (request, response) might make practical sense for `HTTP/2`.

I think we are missing events right now such as:
- OnCreateFailed - OS can't create a handle for HTTP request (out of resources or alike)
- OnConnecting - that should have an opportunity to insert a callback for cert pinning
- OnConnectFailed - that should have ability to indicate that TLS cert validation failed; or if it's a timeout, etc.

Something like this --- borrowed from Apache License 2.0 code :)

```cpp
    ///
    /// API methods:
    /// 1. create
    /// 2. send
    /// 3. cancel (abort)
    /// 4. destroy
    ///
    /// HTTP state machine events:
    /// create         -> [OnCreated    | OnCreateFailed  ]
    /// OnConnecting   -> [OnConnected  | OnConnectFailed ]
    /// send           -> [OnSending                      ]
    /// OnSending      -> [OnSendFailed | OnResponse      ]
    /// destroy        -> [OnDestroy                      ]
    /// cancel         -> [OnCanceled*?                   ]
    ///
    typedef enum
    {
        // create states:
        OnCreated,
        OnCreateFailed,
        // connecting upon creation
        OnConnecting,
        OnConnectFailed,
        OnConnected,
        // sending after connection established
        OnSendFailed,
        OnSending,
        // receiving response
        OnResponse,
        // no response is covered by OnSendFailed
        // destroy after success or cancellation
        OnDestroy
    } HttpStateEvent;
```

Reply:
Good point. Have renamed `ResponseHandler` to `EventHandler` with below methods now:
```cpp
class EventHandler
{
public:
   // called when http response is received.
  virtual void OnResponse(Response &) noexcept = 0;

  // called in case of any error, and no response is received.
  virtual void OnError(SessionState, nostd::string_view &) noexcept = 0;

// called during SSL Handshake for SSL Certificate validation. If certificate is empty, certificate are read from default ca cert path.
  virtual void OnSslValidation(nostd::function_ref<SSLCertificate &()>) noexcept {}

  virtual ~EventHandler() = default;
};
```